test/micro/org/openjdk/bench/jdk/incubator/foreign/AllocatorsForLongRun.java

```diff
@@ -0,0 +1,169 @@
+      }
+    }
+
+    POOL_MAX_SIZE = new int[Long.SIZE + 1];
+    // Use larger sizes for pool, so it will not get exhausted
+    Arrays.fill(POOL_MAX_SIZE, 256);
+  }
+
+  @Param({"1", "16", "200"})
+  public int allocations;
+
+  private MemorySegmentPool pool = new MemorySegmentPool(POOL_MAX_SIZE, ResourceScope.globalScope());
+  private MemorySegmentPool poolEmpty = new MemorySegmentPool(new int[Long.SIZE], ResourceScope.globalScope());
+
+  private int i;
+  @Setup
+  public void setup() {
+    i = 0;
+    // Preallocate pool
+    for (int j = 0; j <= 24; j++) {
+      pool.getSegmentEntryBySize(1L << j, 1).close();
+    }
+  }
+
+  @Benchmark
+  public void arena() {
+    try (var scope = ResourceScope.newConfinedScope()) {
+      final var allocator = SegmentAllocator.arenaAllocator(scope);
+      for (int j = 0; j < allocations; j++) {
+        final var segment = allocator.allocate(sizes[i]);
+        readSegment(segment);
+        next();
+      }
+    }
+  }
+
+  @Benchmark
+  public void pool_allocator() {
+    try (var scope = ResourceScope.newConfinedScope()) {
+      final var allocator = pool.allocatorForScope(scope);
+      for (int j = 0; j < allocations; j++) {
+        final var segment = allocator.allocate(sizes[i]);
+        readSegment(segment);
+        next();
+      }
+    }
+  }
+
+  @Benchmark
+  public void pool_allocator_exhausted() {
+    try (var scope = ResourceScope.newConfinedScope()) {
+      final var allocator = poolEmpty.allocatorForScope(scope);
+      for (int j = 0; j < allocations; j++) {
+        final var segment = allocator.allocate(sizes[i]);
+        readSegment(segment);
+        next();
+      }
+    }
+  }
+
+  @Benchmark
+//  @CompilerControl(CompilerControl.Mode.PRINT)
+  public void pool_direct() {
+    List<MemoryPoolSegment> pooledSegments = new ArrayList<>(allocations);
+    for (int j = 0; j < allocations; j++) {
+      var s = pool.getSegmentEntryBySize(sizes[i], 1);
+      pooledSegments.add(s);
+      readSegment(s.memorySegment());
+      next();
+    }
+    pooledSegments.forEach(MemoryPoolSegment::close);
+  }
+  @Benchmark
+//  @CompilerControl(CompilerControl.Mode.PRINT)
+  public void malloc_free() {
+    List<MemoryAddress> allocatedAddresses = new ArrayList<>(allocations);
+    for (int j = 0; j < allocations; j++) {
+      var size = sizes[i];
+      var a = CLinker.allocateMemory(size);
+      var s = a.asSegment(size, ResourceScope.globalScope());
+      allocatedAddresses.add(a);
+      readSegment(s);
+      next();
+    }
+    allocatedAddresses.forEach(CLinker::freeMemory);
+  }
+
+  /**
+   * Do read to avoid situation allocator will allocate not mapped memory.
+   */
+  private void readSegment(MemorySegment s) {
+    final var size = (int) s.byteSize();
+    for (int idx = 0; idx <  size; idx += 1024) {
+//      MemoryAccess.setByteAtOffset(s, l, (byte)0);
+      BYTE.set(s, 0, (byte) 1);
```

Comment:
You need to cast the offset (idx?) back to a long when you call BYTE.set

Reply:
That said, IMHO, in the new benchmark, the cost of reading dominates everything else. It's not an issue with the memory access API either - replacing read with Unsafe::putInt it still slow. Honestly, since you are after measuring allocation performances, I'm not sure if `readSegment` even belongs here. The more allocations you have, with bigger sized segments, the more read will dominate anything else, to the point of rendering the benchmark not very useful to measure what you want to measure. In fact, even the `strlen` call was an issue in the previous benchmark - as the size of string grows, the time spent in copying data, and strlen grows as well, which then masks the allocation costs. Here you have a similar situation, I think.

Reply:
That's a pity JMH does not allow sub measurements

Reply:
I think that issue is related how JVM shapes code. If I'll write something like this 
readSegment(s.memoryAddress.asSegment(size, scope));
so I'll allocate a new segment based on address than the read performance goes up, and it's same as malloc version

Reply:
See my other code comment - I think the issue here might be that the VM can't clearly see the segment size, so counted loop optimizations are not kicking in.

Reply:
Yeah - I confirm that if you add another getter for `MemoryPoolSegment::memorySegment`, like this:

```
@ForceInline
    public MemorySegment memorySegment(long size) {
      return memoryAddress.asSegment(size, scope);
    }
```
    
Then performance improve considerably. This smells like an inlining issue - e.g. the `size` value that is passed by the code to `getSegmentEntryBySize` is not recognized as the size of the segment - C2 doesn't seem to be able to see through it. There is too much control flow inside `getSegmentEntryBySize` - the trick is to create a "fresh" segment as close as possible to the application code.

The safe version of the API already does this, in fact suffers less from this issue (it's slower than malloc for smaller allocation sizes, but that has always been the case - but it's faster than the direct version!).